app/assets/javascripts/angular/directives/graph_chart.js

```diff
@@ -54,7 +54,10 @@ angular.module("Prometheus.directives").directive('graphChart', ["$location", "W
         var graphHeight = WidgetHeightCalculator(element[0], scope.aspectRatio);
         $el.css('height', graphHeight);
 
+        var seriesState = {};
         if (rsGraph) {
+          // Create a unique name using the series labels to store legend state for newly drawn graph.
+          rsGraph.series.forEach(function(s) { seriesState[renderLabels(s.labels)] = s.disabled });
```

Comment:
Regarding my first paragraph, you might want to go for the first suggestion, as it will keep the code simpler (just save the disabled ones - you don't need to check first if any are disabled).